compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -316,7 +316,17 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
 
         let infcx = self.selcx.infcx();
 
-        if obligation.predicate.has_projections() {
+        // Eagerly normalize obligations, except for `WellFormed` obligations. These
+        // must not be normalized. Normalizing these would allow for projections and
+        // constants which are not well formed if they normalize to something which is.
+        //
+        // See #91068 for more details on how this can be unsound.
+        if obligation.predicate.has_projections()
+            && !matches!(
+                obligation.predicate.kind().skip_binder(),
+                ty::PredicateKind::WellFormed(_)
+            )
+        {
```

Comment:
Does this change stand on its own?

Reply:
moved into #99217, also required for the unevaluated constants changes but this feels important enough that I want to land it quickly.